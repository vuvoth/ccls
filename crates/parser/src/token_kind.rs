use logos::Logos;
use serde::Serialize;

#[derive(Logos, Debug, PartialEq, Clone, Copy, Eq, PartialOrd, Ord, Hash, Serialize)]
#[allow(non_camel_case_types)]
#[repr(u16)]
pub enum TokenKind {
    // Error
    #[error]
    Error = 0,
    // Comments
    #[regex(r"//[^\r\n]*")]
    CommentLine,
    #[token("/*")]
    CommentBlockOpen,
    #[token("*/")]
    CommentBlockClose,
    // Trivial
    #[regex("[ \t]+")]
    WhiteSpace,
    #[regex(r"\r?\n")]
    EndLine,
    // Pragma
    Pragma,
    #[token("pragma")]
    PragmaKw,
    #[token("circom")]
    Circom,
    #[regex("2.[0-9].[0-9]")]
    Version,
    // Literals
    #[regex("[0-9]+")]
    Number,
    #[regex("[$_]*[a-zA-Z][a-zA-Z0-9_$]*")]
    Identifier,
    #[regex(r#""[^"]*""#)]
    CircomString,
    // Brackets
    #[token("(")]
    LParen,
    #[token(")")]
    RParen,
    #[token("{")]
    LCurly,
    #[token("}")]
    RCurly,
    #[token("[")]
    LBracket,
    #[token("]")]
    RBracket,
    // Punctuation
    #[token(";")]
    Semicolon,
    #[token(",")]
    Comma,
    #[token(".")]
    Dot,
    // Boolean operators
    #[token("&&")]
    BoolAnd,
    #[token("||")]
    BoolOr,
    #[token("!")]
    Not,
    // Relational operators
    #[token("==")]
    Equal,
    #[token("!=")]
    NotEqual,
    #[token("<")]
    LessThan,
    #[token(">")]
    GreaterThan,
    #[token("<=")]
    LessThanAndEqual,
    #[token(">=")]
    GreaterThanAndEqual,
    // Arithmetic operators
    #[token("+")]
    Add,
    #[token("-")]
    Sub,
    #[token("*")]
    Mul,
    #[token("**")]
    Power,
    #[token("/")]
    Div,
    #[token("\\")]
    IntDiv,
    #[token("%")]
    Mod,
    // Combined arithmetic assignment
    #[token("+=")]
    AddAssign,
    #[token("-=")]
    SubAssign,
    #[token("*=")]
    MulAssign,
    #[token("**=")]
    PowerAssign,
    #[token("/=")]
    DivAssign,
    #[token(r"\=")]
    IntDivAssign,
    #[token("%=")]
    ModAssign,
    #[token("++")]
    UnitInc,
    #[token("--")]
    UnitDec,
    // Bitwise operators
    #[token("&")]
    BitAnd,
    #[token("|")]
    BitOr,
    #[token("~")]
    BitNot,
    #[token("^")]
    BitXor,
    #[token(">>")]
    ShiftR,
    #[token("<<")]
    ShiftL,
    // Combined bitwise assignments
    #[token("&=")]
    BitAndAssign,
    #[token("|=")]
    BitOrAssign,
    #[token("~=")]
    BitNotAssign,
    #[token("^=")]
    BitXorAssign,
    #[token(">>=")]
    ShiftRAssign,
    #[token("<<=")]
    ShiftLAssign,
    // Assign
    #[token("=")]
    Assign,
    #[token("===")]
    EqualSignal,
    #[token("-->")]
    LAssignSignal,
    #[token("==>")]
    LAssignContraintSignal,
    #[token("<--")]
    RAssignSignal,
    #[token("<==")]
    RAssignConstraintSignal,
    // Conditional expressions
    #[token("?")]
    MarkQuestion,
    #[token(":")]
    Colon,
    // Keywords
    #[token("template")]
    TemplateKw,
    #[token("function")]
    FunctionKw,
    #[token("component")]
    ComponentKw,
    #[token("main")]
    MainKw,
    #[token("public")]
    PublicKw,
    #[token("signal")]
    SignalKw,
    #[token("var")]
    VarKw,
    #[token("include")]
    IncludeKw,
    #[token("input")]
    InputKw,
    #[token("output")]
    OutputKw,
    #[token("log")]
    LogKw,
    // Statement keywords
    #[token("if")]
    IfKw,
    #[token("else")]
    ElseKw,
    #[token("for")]
    ForKw,
    #[token("while")]
    WhileKw,
    #[token("return")]
    ReturnKw,
    #[token("assert")]
    AssertKw,
    // Statements
    IfStatement,
    AssertStatement,
    LogStatement,
    ReturnStatement,
    AssignStatement,
    ForLoop,
    WhileLoop,
    // Program
    CircomProgram,
    // Function
    FunctionDef,
    FunctionName,
    // Template
    TemplateDef,
    TemplateName,
    // Signal
    SignalDecl,
    InputSignalDecl,
    OutputSignalDecl,
    SignalHeader,
    SignalIdentifier,
    // Variable
    VarDecl,
    VarIdentifier,
    // Component
    ComponentDecl,
    ComponentCall,
    ComponentIdentifier,
    SignalOfComponent,
    // Expression
    ExpressionAtom,
    Expression,
    // Complex token kind
    Block,
    ParameterList,
    Call,
    TenaryConditional,
    Condition,
    Statement,
    StatementList,
    ArrayQuery,
    ParserError,
    BlockComment,
    EOF,
    ROOT,
    __LAST,
}

impl From<u16> for TokenKind {
    #[inline]
    fn from(d: u16) -> TokenKind {
        assert!(d <= (TokenKind::__LAST as u16));
        unsafe { std::mem::transmute::<u16, TokenKind>(d) }
    }
}

impl From<rowan::SyntaxKind> for TokenKind {
    fn from(value: rowan::SyntaxKind) -> Self {
        match value {
            rowan::SyntaxKind(id) => TokenKind::from(id),
        }
    }
}

impl From<TokenKind> for u16 {
    #[inline]
    fn from(k: TokenKind) -> u16 {
        k as u16
    }
}

impl From<TokenKind> for rowan::SyntaxKind {
    fn from(kind: TokenKind) -> Self {
        Self(kind as u16)
    }
}

impl TokenKind {
    // a + 10 --> a and 10 are literals
    pub fn is_literal(self) -> bool {
        matches!(self, Self::Number | Self::Identifier)
    }

    // these tokens have the lowest priority
    // <identifier1> infix_operator <identifier2>
    // eg: a + b --> + is an infix token
    pub fn infix(self) -> Option<(u16, u16)> {
        match self {
            // arithmetic operators
            Self::Power => Some((99, 100)),
            Self::Mul | Self::Div | Self::IntDiv | Self::Mod => Some((94, 95)),
            Self::Add | Self::Sub => Some((89, 90)),
            // shift bitwise operators
            Self::ShiftL | Self::ShiftR => Some((84, 85)),
            // relational operators
            Self::LessThan
            | Self::GreaterThan
            | Self::LessThanAndEqual
            | Self::GreaterThanAndEqual => Some((79, 80)),
            Self::Equal | Self::NotEqual => Some((74, 75)),
            // other bitwise operators
            Self::BitAnd => Some((69, 70)),
            Self::BitXor => Some((64, 65)), // exclusive or
            Self::BitOr => Some((59, 60)),
            // boolean operators
            Self::BoolAnd => Some((54, 55)),
            Self::BoolOr => Some((49, 50)),
            // ----------
            // TODO: how about conditional operation ( ? : )
            // associativity: right to left [ a ? b : c --> ??? ]

            // ----------
            // associativity: right to left [ a = b = c --> a = (b = c) ]
            // DO NOT CONSIDER ASSIGMENT OPERATORS AS INFIX TOKENS
            /*
            // assignment operators
            Self::Assign
            // signal assigment operators
            | Self::EqualSignal
            | Self::LAssignSignal
            | Self::LAssignContraintSignal
            | Self::RAssignSignal
            | Self::RAssignConstraintSignal
            // bitwise asignment operators
            | Self::BitOrAssign
            | Self::BitXorAssign
            | Self::BitAndAssign
            | Self::ShiftLAssign
            | Self::ShiftRAssign
            // arithmetic asignament operators
            | Self::AddAssign
            | Self::SubAssign
            | Self::MulAssign
            | Self::DivAssign
            | Self::IntDivAssign
            | Self::ModAssign
            | Self::PowerAssign => Some((44, 45)),
            */
            // TODO: how about comma (expression separator)
            Self::Comma => Some((39, 40)),
            // not an infix operator
            _ => None,
        }
    }

    // priority: post > pre > in
    // associativity: right to left [ --!a --> --(!a) ]
    // prefix_operator <literal>
    // eg: -10, !a, ++a, --a
    pub fn prefix(self) -> Option<u16> {
        match self {
            Self::UnitDec | Self::UnitInc | Self::Sub | Self::Add | Self::Not | Self::BitNot => {
                Some(200)
            }

            _ => None,
        }
    }

    // these tokens have the highest priority
    // <literal> postfix_operator
    // eg: a[10], b++, c.att1
    pub fn postfix(self) -> Option<u16> {
        match self {
            Self::LParen // function call
            | Self::LBracket // array subscript
            | Self::Dot // attribute access
            | Self::UnitDec | Self::UnitInc => Some(300),

            _ => None,
        }
    }

    pub fn is_declaration_kw(self) -> bool {
        matches!(self, Self::VarKw | Self::ComponentKw | Self::SignalKw)
    }

    pub fn is_assign_token(self) -> bool {
        matches!(
            self,
            Self::Assign
            // signal assigment operators
            | Self::EqualSignal
            | Self::LAssignSignal
            | Self::LAssignContraintSignal
            | Self::RAssignSignal
            | Self::RAssignConstraintSignal
            // bitwise asignment operators
            | Self::BitOrAssign
            | Self::BitXorAssign
            | Self::BitAndAssign
            | Self::ShiftLAssign
            | Self::ShiftRAssign
            // arithmetic asignament operators
            | Self::AddAssign
            | Self::SubAssign
            | Self::MulAssign
            | Self::DivAssign
            | Self::IntDivAssign
            | Self::ModAssign
            | Self::PowerAssign // unit inc/dec
                                // | Self::UnitInc
                                // | Self::UnitDec
        )
    }

    pub fn is_inline_assign_signal(self) -> bool {
        matches!(
            self,
            Self::Assign | Self::RAssignSignal | Self::RAssignConstraintSignal
        )
    }

    pub fn is_var_assign(self) -> bool {
        matches!(self, Self::Assign)
    }

    pub fn is_trivial(self) -> bool {
        matches!(
            self,
            Self::WhiteSpace | Self::EndLine | Self::CommentLine | Self::BlockComment | Self::Error
        )
    }
}
