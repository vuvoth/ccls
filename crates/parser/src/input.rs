use std::ops::Range;

use logos::Lexer;

use serde::Serialize;

use crate::token_kind::TokenKind;

#[derive(Debug, PartialEq, Serialize)]
pub struct Input<'a> {
    kind: Vec<TokenKind>,
    source: &'a str,
    position: Vec<Range<usize>>,
}

impl<'a> Input<'a> {
    pub fn new(source: &'a str) -> Self {
        let mut input = Input {
            source,
            kind: Vec::new(),
            position: Vec::new(),
        };

        let mut lex = Lexer::<TokenKind>::new(source);

        while let Some(tk) = lex.next() {
            if tk == TokenKind::CommentBlockOpen {
                let mut closed = false;
                let mut join_span = lex.span();
                while let Some(t) = lex.next() {
                    join_span.end = lex.span().end;
                    if t == TokenKind::CommentBlockClose {
                        closed = true;
                        break;
                    }
                }

                if closed {
                    input.kind.push(TokenKind::BlockComment);
                } else {
                    input.kind.push(TokenKind::Error);
                }
                input.position.push(join_span);
            } else {
                input.kind.push(tk);
                input.position.push(lex.span());
            }
        }

        input
    }

    pub fn token_value(&self, index: usize) -> Option<&'a str> {
        if index < self.kind.len() {
            Some(&self.source[self.position[index].start..self.position[index].end])
        } else {
            // return None for out of bound index
            None
        }
    }

    pub fn kind_of(&self, index: usize) -> TokenKind {
        if index < self.kind.len() {
            self.kind[index]
        } else {
            TokenKind::EOF
        }
    }

    pub fn position_of(&self, index: usize) -> Option<Range<usize>> {
        if index < self.kind.len() {
            Some(self.position[index].clone())
        } else {
            // return error for out of bound index
            None
        }
    }

    pub fn size(&self) -> usize {
        self.kind.len()
    }
}

#[cfg(test)]
mod tests {
    use super::Input;

    fn test(source: &str, snapshot_name: &str) {
        let input = Input::new(&source);

        insta::assert_yaml_snapshot!(snapshot_name, input);
    }

    #[test]
    fn test_comment_block() {
        let source = r#"
        /*a + b == 10*/
        a + 10
    "#;
        test(source, "test_comment_block");
    }

    #[test]
    fn test_comment_error() {
        let source = r#"
        pragma 2.1.1;
        /*a + b == 10*
        a + 10
        template

        /*
    "#;
        test(source, "test_comment_error");
    }

    #[test]
    fn test_pragma() {
        let source = r#"
        /* test pragma token kinds */

    pragma circom 2.0.0;

    "#;
        test(source, "test_pragma");
    }

    #[test]
    fn test_function() {
        let source = r#"
    function nbits(a) {
        var n = 1;
        var r = 0;
        while (n-1<a) {
            r++;
            n *= 2;
        }
        return r;
    }"#;
        test(source, "test_function");
    }

    #[test]
    fn test_operators() {
        let source = r#"
        ({[]})
        ;.,:
        && &
        || |
        != !
        === == =
        --> ==>
        <-- <==
        <= <
        >= >
        ++ += +
        -- -= -
        **= **
        * *=
        / /=
        \ \=
        % %=
        ^ ^=
        ~ ~=
        >> >>=
        << <<=
        & &=
        | |=
    }"#;
        test(source, "test_operators");
    }
}
