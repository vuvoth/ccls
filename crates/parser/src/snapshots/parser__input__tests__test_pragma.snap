---
source: crates/parser/src/input.rs
expression: input
---
kind:
  - EndLine
  - WhiteSpace
  - BlockComment
  - EndLine
  - EndLine
  - WhiteSpace
  - PragmaKw
  - WhiteSpace
  - Circom
  - WhiteSpace
  - Version
  - Semicolon
  - EndLine
  - EndLine
  - WhiteSpace
source: "\n        /* test pragma token kinds */\n\n    pragma circom 2.0.0;\n\n    "
position:
  - start: 0
    end: 1
  - start: 1
    end: 9
  - start: 9
    end: 38
  - start: 38
    end: 39
  - start: 39
    end: 40
  - start: 40
    end: 44
  - start: 44
    end: 50
  - start: 50
    end: 51
  - start: 51
    end: 57
  - start: 57
    end: 58
  - start: 58
    end: 63
  - start: 63
    end: 64
  - start: 64
    end: 65
  - start: 65
    end: 66
  - start: 66
    end: 70
